thunder/torch/__init__.py

```diff
@@ -5590,7 +5590,12 @@ def backward_autograd_function_apply(
     id="torch.amp.autocast_mode._enter_autocast",
     tags=(prims.OpTags.DONT_DCE, prims.OpTags.CTX_MANAGER_ENTER_EXIT_OP),
 )
-def autocast_enter(device_type, dtype=None, enabled=True):
+def autocast_enter(device_type, dtype=None, enabled=True, _unused_cache_enabled=True):
+    # We may receive device_type=cuda:0
+    # PyTorch applies autocast irrespective of device index.
+    # So, here we grab the device_type from the string.
+    device_type, unused_deviceno = devices._device_from_string_helper(device_type)
+    device_type = devices.devicetype_string(device_type)
```

Comment:
As per [docs](https://pytorch.org/docs/stable/amp.html#torch.autocast) `torch.autocast` takes `device_type`, but in practice it does take device + index eg. `cuda:0` and ignores the `index` part. Not sure if it was intended but for now, we mimic the behaviour.

Eg.

```python
import torch

device = 'cuda:1'
dtype = torch.float

x = torch.rand(2, 2, device=device, dtype=dtype, requires_grad=True)

def func(x):
    with torch.autocast("cuda:0"):  # NOTE: This is cuda:0
        return torch.matmul(x, x)

o = func(x)
print(o.dtype)  # torch.float16

```